src/output/plugins/visualization/SoundAnalysis.cxx

```diff
@@ -0,0 +1,394 @@
+	idx_hi = (size_t) ceil(freq_hi * (float)num_samples / samples_per_sec);
+
+	float num_octaves = log(freq_hi/freq_lo) / 0.69314718f;
+
+	float freq_mids	 = freq_lo * pow(2.0f, num_octaves / 3.0f);
+	float freq_trebs = freq_lo * pow(2.0f, 2.0f * num_octaves / 3.0f);
+
+	idx_mids  = ns *  freq_mids / samples_per_sec;
+	idx_trebs = ns * freq_trebs / samples_per_sec;
+}
+
+bool
+Visualization::SoundAnalysis::Update(SoundInfoCache::Time t) noexcept
+{
+	FmtDebug(d_sound_analysis, "SoundAnalysis::update(tid: {}), time {}us, # samp: {}, "
+			 "buffer size: {}", gettid(), t.count(), num_samples, pcache->Size());
+
+	if (!pcache->GetByTime(num_samples, t, buf.get(), cbbuf)) {
+		FmtWarning(d_sound_analysis, "Failed to get samples by time for sound analysis "
+				   "({} samples requested, at time {}us for buf size {}).", num_samples, 
+				   t.count()/1000, cbbuf);
+		return false;
+	}
+
+	/* Copy the raw PCM data from `buf` into `in`. I hate this, but we need to
+	 * convert the input data from `uint16_t` (or whatever) to `float`
+	 * regardless. We could, of course, do the conversion when the PCM data is
+	 * added to the cache, but since I anticipate processing far fewer samples
+	 * than I store, I expect this to be more efficient (both in terms of time &
+	 * space).
+	 * 
+	 * Since we have to do the copy anyway, let's convert from interleaved to
+	 * sequential (i.e. all samples for the first channel laid-out contiguously,
+	 * followed by all from the second, and so forth). */
+	typedef IntegerToFloatSampleConvert<SampleFormat::S8> s8_cvt_t;
+	typedef IntegerToFloatSampleConvert<SampleFormat::S16> s16_cvt_t;
+	typedef IntegerToFloatSampleConvert<SampleFormat::S32> s32_cvt_t;
+	typedef IntegerToFloatSampleConvert<SampleFormat::S24_P32> s24_p32_t;
+
+	for (size_t i = 0; i < num_samples; ++i) {
+		for (size_t j = 0; j < num_channels; ++j) {
+			/* `buf` index: i * num_channels + j
+			 * `in` index: j * num_samples + i */
+			float x;
+			switch (audio_format.format) {
+			case SampleFormat::S8:
+				x = s8_cvt_t::Convert(*(char*)buf[i * num_channels + j]);
+				break;
+			case SampleFormat::S16:
+				x = s16_cvt_t::Convert(*(int16_t*) (buf.get() + 2 * (i * num_channels + j)));
+				break;
+			case SampleFormat::S32:
+				// x = (float)*(int32_t*)(buf.get() + 4 * (i * num_channels + j));
+				x = s32_cvt_t::Convert(*(int32_t*)(buf.get() + 4 * (i * num_channels + j)));
+				break;
+			case SampleFormat::FLOAT:
+				x = *(float*)(buf.get() + 4 * (i * num_channels + j));
+				break;
+			case SampleFormat::S24_P32:
+				/* signed 24 bit integer samples, packed in 32 bit integers
+				 * (the most significant byte is filled with the sign bit) */
+				x = s24_p32_t::Convert(*(int32_t *)(buf.get() + 4 * (i * num_channels + j)));
+				break;
+			default:
+				assert(false);
+			}
+			in.get()[j * num_samples + i] = x;
+		}
+	}
+
+	fftwf_execute(plan);
+
+	size_t max_coeffs_idx = num_samples/2;
+
+	for (uint8_t c = 0; c < num_channels; ++c) {
```

Comment:
Using `uint8_t` here instead of `unsigned` has no effect at best, and can add overhead and code bloat. It's a harmful micro-optimization.

Reply:
Wasn't intended as an optimization or any sort; just chose the type to match that of `num_channels`. Changed, but can you say more about how making `c` a `uint8_t` could add overhead & code bloat?